typescript/cognito-idp/README.md

```diff
@@ -0,0 +1,71 @@
+# A web site and REST API with Cognito authentication using the Facebook identity pool provider
+
+The example will show you how to create the following:
+
+- A single-page app hosted by S3 and CloudFront
+- A REST API that uses Cognito for authentication
+- Integration of Facebook as an identity provider
+
+It also demonstrates a somewhat opinionated way to organize your lambda functions and test them.
+
+## Prerequisites
+
+There is a bit of setup required before you can deploy this stack.
+
+- A domain that you control with Route53 in your development account
+    - https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/getting-started.html
+    - Be careful about Cognito reserved domains.
+        - You can't have 'aws' or 'cognito' anywhere in the domain name
+- A certicate that covers the web site and the API, created with ACM
+    - For example, you own "example.com"
+    - You have a certificate for "myapp.example.com" and "myapi.example.com"
+    - https://docs.aws.amazon.com/acm/latest/userguide/gs.html
+- A Facebook Developer account
+    - Create an app and put the app id into your .env
+    - Store the app secret in AWS Secrets Manager as facebook_app_secret
+    - https://developers.facebook.com/docs/facebook-login/
+    - https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-social-idp.html
+    - (Don't follow the separate guide for integrating Facebook with Identity Pools)
+
+## Environment Variables
+
+Put the following into a .env file in the root folder for this example, replacing each placeholder value with your specific environment values.
+
+```
+AWS_REGION=us-east-1
+AWS_ACCOUNT=123456789123
+WEB_DOMAIN=myapp.example.com
+API_DOMAIN=myapi.example.com
+WEB_CERTIFICATE_ARN=
+API_CERTIFICATE_ARN=
+FACEBOOK_APP_ID=
+FACBOOK_VERSION=v7.0
+FACEBOOK_SECRET_ARN=
+COGNITO_REDIRECT_URI=
+COGNITO_DOMAIN_PREFIX=
+COGNITO_REGION=
+```
```

Comment:
Yes I can do that, but maybe I should wait until we decide the best way to handle environment variables.

Reply:
Would these be best located in the CDK context variable file? Instead of external environment variables?

Reply:
> maybe I should wait until we decide the best way to handle environment variables.

What is the pending decision you're referring to?

Reply:
I suppose you're referring to the comment here - https://github.com/aws-samples/aws-cdk-examples/pull/298#discussion_r436741895

Reply:
Yes. When it comes down to it, it doesn't really make sense to synthesize this stack without the environment variables being set. There are no defaults that would result in a successful deployment. So at some point, we need to have a place to put environment variables. cdk.json won't work since that gets committed. Specifying context variables on the command line just moves the problem to a different place. I would probably convert that long list of arguments to an npm script, which would still require something like cross-env to refer to a .env file.

Reply:
Our customers would check them into git as well though. There’s nothing inherently secret about any of these values.

Reply:
How would that work on a team, though? Each developer would need their own values for each of those. And then dev, beta, gamma, prod environments would all need their own values.

Reply:
Our stance in general is that “degrees of freedom” should generally be expressed through code APIs and fully resolved during synthesis so that synthesis is “pure” and remains deterministic. The invariant is that given a commit in the repo, the resulting cloud assembly will always be exactly the same for any call of “synth”. Since CDK apps are executed during build, this is a similar invariant that compilers have and we have assumptions in our framework that rely on this, and generally a best practice from an operational standpoint. That’s also the reason we discourage querying external APIs during synthesis.

We should also consider the fact that this is an “official” example. As such it should not encourage anti patterns.

Having said all that, the development use case is slightly different, because as @ericzbeard noted, each developer would have their own environment and those values will be different for each developer, in which case I do think that environment variables are a viable approach, but there are numerous other solutions that can be used and perhaps will be easier to work with.

For example, the code can read those values from a simple json file. By default it can read the “production” file from the project directory and an environment variable can be used to point it to a different file, for each developer.

Here is what I propose:

Define a `MyStackProps` interface with all these attributes and read its contents from a json file. Commit a file to source control with dummy “production” values (normally these would be the actual production values but since this is a sample, we will just put some dummy values).

Allow the developer to override the file’s location through an environment variable. 

Sketch:

```ts
const configFile = process.env.CONFIG_FILE ?? “./config.json”;
new MyStack(app, “stack”, require(configFile));
```

What do you guys think? We know we need to publish more guidance on this topic and we even have an rfc tracking item for that, but never got around to it. @ericzbeard help!

Reply:
I like it.

Reply:
This change won't be too hard to implement and I suppose it's more idiomatic for CDK, but I'm not sure if it's what developers will expect. A .env file is an extremely common and well-understood pattern. 

I'm also not sure about checking in a "production" config file. I have always seen checking in config values for any environment as an anti-pattern. Especially if eventually the checked in config file has actual production values and not placeholders - that could potentially cause outages if developers have privileges to the production system. 

I often create a sample config file for reference and check that in, or simply document it in a README. I suppose if we want the project to build and synth with no additional action from a user (and for the purposes of an automated CI server) we don't have a choice, but I worry about a user cloning the project, building, and attempting to deploy it without creating their own config file.  That will definitely fail and might be confusing. As it is, synth fails with a clear message that you are missing a necessary environment variable. How do we solve that problem?

Once we figure this out, it should definitely be built into the CLI somehow, probably using context variables. We already have cdk.json but that gets checked in. Maybe we should have a cdk.local.json that gets picked up automatically if present and is added to the .gitignore file when creating a new app?

Reply:
For the time being, I committed a change to move all environment variable loading to the file in bin/ where we create the app. The stack no longer has any direct references to .env, and uses props that are passed in. This still leaves several open questions from above.